Source/control.cpp

```diff
@@ -267,11 +267,33 @@ void PrintInfo(const Surface &out)
 
 	auto infoStringLines = (int)std::count(InfoString.str().begin(), InfoString.str().end(), '\n');
 
-	Rectangle line { GetMainPanel().position + Displacement { 177, LineStart[infoStringLines + pnumlines] }, { 288, 12 } };
+	int panelLines = pnumlines + infoStringLines;
+
+	std::string infoStringOut = std::string(InfoString.str());
+
+	if (panelLines >= 4) {
+		panelLines = 3;
+		pnumlines = panelLines - infoStringLines;
+		Log("PrintInfo unable to render everything - not enough lines");
+	}
+
+	Rectangle line { GetMainPanel().position + Displacement { 177, LineStart[panelLines] }, { 288, 12 } };
 
 	if (!InfoString.empty()) {
-		int lineHeight = LineHeights[pnumlines + infoStringLines];
-		DrawString(out, InfoString, line, InfoColor | UiFlags::AlignCenter | UiFlags::KerningFitSpacing, 2, lineHeight);
+		const int lineHeight = LineHeights[panelLines];
+		if (infoStringLines >= 4) {
+			infoStringLines = 3;

```

Comment:
Rather than doing all these memory allocations on every frame, how about we only do this when `InfoString` is changed?